### PR TITLE
Rewrite eval.sh to python and add a json output for evaluation metrics

### LINKED ISSUE #337: Evaluation results should be saved as JSON or YAML
Right now they are formatted as `f"{bleu}\n{chrf}"` which is not very helpful. They should be output to include all of the information deduced by:

[utils/tc_marian_logs.py](https://github.com/mozilla/firefox-translations-training/pull/335/files#diff-b0e7241806fe6393f3d20c5f0a4f6b31a51c79f155f003f458eb69f563f25262) (See pull #335).

### PR DESCRIPTION
**Pull request by gregtatum**
This is a follow-up to #393, and fixes #337. It also converts eval.sh into eval.py and simplifies the taskgraph.

### CODE REVIEW

File: pipeline/eval/eval.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+    """
+    command_dedented = dedent(command)
+
+    # Remove comments and whitespace.
+    lines = [
+        line.strip() for line in command_dedented.split("\n") if line and not line.startswith("#")
+    ]
+    command = " \\\n".join(lines)
+
+    print("-----------------Running bash in one line--------------")
```
Comment (gregtatum) [inline pipeline/eval/eval.py:64]: This should eventually be put into a shared location like `pipeline/utils`.

**Comment by gregtatum** [comment]
The eval step was cached, but here it is in CI:

https://firefox-ci-tc.services.mozilla.com/tasks/groups/O3ih2h9tTt2CLbgsN5cZXw

File: pipeline/eval/eval.py (hunk 1 above)
Comment (gregtatum) [reply pipeline/eval/eval.py:64]: Here is some example output:

```
[task 2024-01-30T15:51:31.305Z] The eval script is configured with the following:
[task 2024-01-30T15:51:31.305Z]  >          artifacts_dir: /home/ubuntu/tasks/task_170662979767690/artifacts
[task 2024-01-30T15:51:31.305Z]  > source_file_compressed: /home/ubuntu/tasks/task_170662979767690/fetches/devtest.en.zst
[task 2024-01-30T15:51:31.305Z]  >            source_file: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.en
[task 2024-01-30T15:51:31.305Z]  >            target_file: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.ru
[task 2024-01-30T15:51:31.305Z]  >        target_ref_file: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.ru.ref
[task 2024-01-30T15:51:31.305Z]  >         marian_decoder: "$MOZ_FETCHES_DIR"/marian-decoder
[task 2024-01-30T15:51:31.305Z]  >        marian_log_file: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.log
[task 2024-01-30T15:51:31.305Z]  >          language_pair: en-ru
[task 2024-01-30T15:51:31.305Z]  >           metrics_file: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.metrics
[task 2024-01-30T15:51:31.305Z]  >           metrics_json: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.metrics.json
[task 2024-01-30T15:51:31.305Z]  >      marian_extra_args: ['--workspace', '12000', '--devices', '0']
[task 2024-01-30T15:51:31.305Z] Ensure that the artifacts directory exists.
[task 2024-01-30T15:51:31.305Z] Save the original target sentences to the artifacts
[task 2024-01-30T15:51:31.305Z] -----------------Running bash in one line--------------
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z]   zstdmt -dc "/home/ubuntu/tasks/task_170662979767690/fetches/devtest.ru.zst" > "/home/ubuntu/tasks/task_170662979767690/artifacts/devtest.ru.ref"
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z] -------------------------------------------------------
[task 2024-01-30T15:51:31.305Z] -----------------Running bash in one line--------------
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z]   # Decompress the source file, e.g. $fetches/wmt09.en.gz
[task 2024-01-30T15:51:31.305Z]   zstdmt -dc "/home/ubuntu/tasks/task_170662979767690/fetches/devtest.en.zst"
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z]   # Tee the source file into the artifacts directory, e.g. $artifacts/wmt09.en
[task 2024-01-30T15:51:31.305Z]   | tee "/home/ubuntu/tasks/task_170662979767690/artifacts/devtest.en"
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z]   # Take the source and pipe it in to be decoded (translated) by Marian.
[task 2024-01-30T15:51:31.305Z]   | "$MOZ_FETCHES_DIR"/marian-decoder
[task 2024-01-30T15:51:31.305Z]       --models /home/ubuntu/tasks/task_170662979767690/fetches/final.model.npz.best-chrf.npz
[task 2024-01-30T15:51:31.305Z]       --config /home/ubuntu/tasks/task_170662979767690/fetches/final.model.npz.best-chrf.npz.decoder.yml
[task 2024-01-30T15:51:31.305Z]       --quiet
[task 2024-01-30T15:51:31.305Z]       --quiet-translation
[task 2024-01-30T15:51:31.305Z]       --log /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.log
[task 2024-01-30T15:51:31.305Z]       --workspace 12000 --devices 0
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z]   # The translations be "tee"ed out to the artifacts, e.g. $artifacts/wmt09.ca
[task 2024-01-30T15:51:31.305Z]   | tee "/home/ubuntu/tasks/task_170662979767690/artifacts/devtest.ru"
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z]   # Take in the translations via stdin and run the evaluation using sacrebleu.
[task 2024-01-30T15:51:31.305Z]   # The translations will be compared to the original corpus via the target_ref_file
[task 2024-01-30T15:51:31.305Z]   | sacrebleu "/home/ubuntu/tasks/task_170662979767690/artifacts/devtest.ru.ref"
[task 2024-01-30T15:51:31.305Z]       --language-pair en-ru
[task 2024-01-30T15:51:31.305Z]       --metrics bleu chrf
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z]   # Finally, tee out the metrics to the json file.
[task 2024-01-30T15:51:31.305Z]   | tee "/home/ubuntu/tasks/task_170662979767690/artifacts/devtest.metrics.json"
[task 2024-01-30T15:51:31.305Z] 
[task 2024-01-30T15:51:31.305Z] -------------------------------------------------------
[task 2024-01-30T15:51:31.305Z] Look up the bleu and chrf from the output files from: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.metrics.json
[task 2024-01-30T15:51:31.305Z] Metrics:  [{'name': 'BLEU', 'score': 0.0, 'signature': 'nrefs:1|case:mixed|eff:no|tok:13a|smooth:exp|version:2.3.1', 'verbose_score': '49.8/0.0/8.3/0.0 (BP = 0.000 ratio = 0.087 hyp_len = 2027 ref_len = 23295)', 'nrefs': '1', 'case': 'mixed', 'eff': 'no', 'tok': '13a', 'smooth': 'exp', 'version': '2.3.1'}, {'name': 'chrF2', 'score': 0.5, 'signature': 'nrefs:1|case:mixed|eff:yes|nc:6|nw:0|space:no|version:2.3.1', 'nrefs': '1', 'case': 'mixed', 'eff': 'yes', 'nc': '6', 'nw': '0', 'space': 'no', 'version': '2.3.1'}]
[task 2024-01-30T15:51:31.305Z] Saving the metrics in the older "text" format: /home/ubuntu/tasks/task_170662979767690/artifacts/devtest.metrics
```

**Review by bhearsum — APPROVED** [review]
I didn't review the new script in detail, but this looks great from a taskgraph standpoint! Thank your for getting rid of the unnecessary `task-defaults`.

File: pipeline/eval/eval.py (hunk 1 above)
Comment (eu9ene) [reply pipeline/eval/eval.py:64]: thanks for logging marian arguments, it's important to see what's happening

**Review by eu9ene — COMMENTED** [review]
It's great to see the evaluation script wrapped in Python! It will allow us to add more metrics like Comet easier. 

Couple of thoughts/suggestions:
1) Can we just use Sacrebleu Python API instead of command line one? https://github.com/mjpost/sacrebleu?tab=readme-ov-file#using-sacrebleu-from-python

2) I have a concern that writing jsons will break compatibility with publishing tools that expect the old file format: https://github.com/mozilla/firefox-translations-training/blob/3d81ca5d11fe4d04151d009226c463417e204f13/tracking/translations_parser/data.py#L60
We have two options: either writing the old file together with json or fixing the code in the publishing to support both. We will want to republish both old and new experiments while we're tuning the dashboards.

**Comment by gregtatum** [comment]
> Can we just use Sacrebleu Python API instead of command line one?

Part of my motivation of keeping the bash was that as far as I can tell, this all streams using pipes. I _think_ if you did python here you would lose some of the benefits of streaming all of the results through. I could be wrong. It would be nice to do some experimentation here on how to efficiently chew through large amounts of data with python vs bash pipelines.

**Comment by gregtatum** [comment]
> It is possible to create a pipe, run a program redirecting its stdout to this pipe and another redirecting its stdin to the same pipe, though I'm not sure how ergonomic it is in Python.

Yeah at this point the bash is simpler for this I think. I've looked at a few ways of doing this, and bash seems easier than python. I think this approach is a bit of a middle of the road implementation where we get the benefits of argument parsing and validation in python, and then the ease of running some commands using bash syntax and the `run_bash_oneliner` script.

It's not immediately obvious to me of the python API even supports streaming here:

https://github.com/mjpost/sacrebleu?tab=readme-ov-file#using-sacrebleu-from-python

**Comment by marco-c** [comment]
If you create the pipe yourself in Python, you can make Marian output to the pipe and call sacrebleu API on the pipe contents as they arrive, basically implementing streaming yourself.

**Pull request merged**